lib/fsm/testhelpers.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fsm
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gravitational/gravity/lib/rpc"
+	"github.com/gravitational/gravity/lib/storage"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/trace"
+	"github.com/pborman/uuid"
+)
+
+func newTestEngine(plan storage.OperationPlan) *testEngine {
+	return &testEngine{plan: plan}
+}
+
+// testEngine is fsm engine used in tests. Keeps its changelog in memory.
+type testEngine struct {
+	plan      storage.OperationPlan
+	changelog storage.PlanChangelog
+}
+
+// GetExecutor returns one of the test executors depending on the specified phase.
+func (t *testEngine) GetExecutor(p ExecutorParams, r Remote) (PhaseExecutor, error) {
+	if strings.HasPrefix(p.Phase.ID, "/init") {
+		return newInit(), nil
+	} else if strings.HasPrefix(p.Phase.ID, "/bootstrap") {
+		return newBootstrap(), nil
+	} else if strings.HasPrefix(p.Phase.ID, "/upgrade") {
+		return newUpgrade(), nil
+	}
+	return nil, trace.BadParameter("unknown phase %q", p.Phase.ID)
+}
+
+// ChangePhaseState records the provided phase state change in the test engine.
+func (t *testEngine) ChangePhaseState(ctx context.Context, ch StateChange) error {
+	t.changelog = append(t.changelog, storage.PlanChange{
+		ID:       uuid.New(),
+		PhaseID:  ch.Phase,
+		NewState: ch.State,
+		Created:  time.Now().UTC(),
```

Comment:
This is using real time on purpose, because when calculating the actual plan state, the most recent changelog entry state wins. Added a comment about this.